### PR TITLE
[wgsl-out] Correct production of `*` and `&` operators.

### LINKED ISSUE #1322: WGSL back end generates incorrect code for address-of expressions
(Bug description revised).

Passing the following WGSL through Naga IR:

```
fn f(i: ptr<function, i32>) {}

fn g() {
  var x = 10;
  f(&x);
}
```

produces the following incorrect output:

```
fn f(i: ptr<function, i32>) {
    return;
}

fn g() {
    var x: i32 = 10;

    f(x);
    return;
}
```

Specifically, the passing of `x` to `f` is ill-typed: the `&` operator has been dropped.

---

Original, incorrect bug report:

Passing the following WGSL through Naga IR:

```
fn f(i: i32) -> i32 {
  var x = i;
  var p = &x;
  return i;
}
```

produces the following incorrect output:

```
fn f(i: i32) -> i32 {
    var x: i32;
    var p: ptr<function, i32>;

    x = i;
    p = x;
    return i;
}
```

Specifically, the assignment of `x` to `p` is ill-typed: the `&` operator has been dropped.

### PR DESCRIPTION
**Pull request by jimblandy**
Fixes #1322.

This still needs tests, so, marking it as a draft, but comments welcome.

### CODE REVIEW

File: src/back/wgsl/writer.rs [hunk 1]
```diff
@@ -22,6 +22,85 @@ enum Attribute {
...
+/// - The plain form of `LocalVariable(x)` is simply `x`, which is a reference
+///   to the local variable's storage.
+///
+/// - The plain form of `GlobalVariable(g)` is simply `g`, which is usually a
+///   reference to the global variable's storage. However, globals in the
+///   `Handle` storage class are immutable, and `GlobalVariable` expressions for
+///   those produce the value directly, not a pointer to it. Such
+///   `GlobalVariable` expressions are `Ordinary`.
+///
+/// - `Access` and `AccessIndex` are `Reference` when their `base` operand is a
```
Comment (kvark) [inline src/back/wgsl/writer.rs:68]: so they just inherit the indirection from the base

File: src/back/wgsl/writer.rs [hunk 2]
```diff
@@ -938,8 +1031,25 @@ impl<W: Write> Writer<W> {
             return Ok(());
         }
 
+        // If the plain form of the expression is not what we need, emit the
+        // operator necessary to correct that.
+        let plain = plain_form_indirection(expr, module, func_ctx);
+        match (requested, plain) {
```
Comment (kvark) [inline src/back/wgsl/writer.rs:1037]: this is really clean, well done!

**Review by kvark — APPROVED** [review]

**Comment by kvark** [comment]
I'd feel good just landing it. This isn't less tested than what we already have :)

**Comment by jimblandy** [comment]
All right, we'll land it as is.

**Pull request merged**